torch/fx/experimental/proxy_tensor.py

```diff
@@ -559,17 +581,19 @@ def __enter__(self):
         m = self.sym_mode.enable(True)
         self._managers.append(m)
         m.__enter__()
-        return super().__enter__()
+        _set_proxy_tensor_mode(self)
```

Comment:
Nyeh, I don't like it. Why doesn't super() work anymore?

Reply:
I can clean this up.

This doesn't work by default anymore because the parent class's `__enter__` calls `_push_on_torch_dispatch_stack(mode)`, which doesn't know to do anything special if the mode is a `Fake`/`Proxy`/`Functional` mode.

(1) I can make the C++ code in `_push_on_torch_dispatch_stack()` know to treat some modes specially

(2) going off your `ModeKey` idea, another thing I could do is shove a `ModeKey` property onto every `TorchDispatchMode`, e.g.:
```
enum:
    Default,
    Functional,
    Proxy,
    Fake
```

Where normal modes get the "default" enum. And then the default implementation of `__enter__` would look something like:
```
def __enter__(self)
torch._C._push_on_torch_dispatch_stack(self, self._mode_key)
```

Reply:
Hmm... it seems to me like the main benefit(s) of inheriting from `TorchDispatchMode` are:

(1) re-using shared infra so that entering/exiting your mode properly modifies our C++ TLS

(2) enforcing uniformity across our modes. Users today can call `torch.utils._python_dispatch._get_current_dispatch_mode_stack()`, and get back a stack of `TorchDispatchMode` invariants.

I agree that (1) has a bit less benefit if the logic to modify our TLS is different for these modes. (2) seems like a reasonable invariant to want to uphold though, what do you think?

And to be fair, I think we can still pretty easily keep our `__enter__` the same for `ProxyTorchDispatchMode` (from above comment)

TODO

Reply:
ok